#Tratativa dos dados



In [25]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [125]:
#Dataset simples
!wget https://raw.githubusercontent.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/main/DADOS_ABERTOS_MEDICAMENTOS.csv

#Dataset com tarjas - INSERIR MANUALMENTE
#!wget https://github.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/blob/main/medTarja.xls

--2022-07-26 00:42:49--  https://raw.githubusercontent.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/main/DADOS_ABERTOS_MEDICAMENTOS.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5903386 (5.6M) [text/plain]
Saving to: ‘DADOS_ABERTOS_MEDICAMENTOS.csv’

DADOS_ABERTOS_MEDIC 100%[===================>]   5.63M  --.-KB/s    in 0.09s   

2022-07-26 00:42:49 (64.7 MB/s) - ‘DADOS_ABERTOS_MEDICAMENTOS.csv’ saved [5903386/5903386]

--2022-07-26 00:42:49--  https://github.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/blob/main/medTarja.xls
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘medTa

In [126]:
pip install xlrd==1.2.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [129]:
df = pd.read_excel("/content/medTarja.xls", engine='xlrd', index_col=None, header=45, skiprows = [0, 45])
df

SUBSTÂNCIA                CNPJ  \
0                      CEFALOTINA SODICA  04.301.884/0001-75   
1                      CEFAZOLINA SÓDICA  04.301.884/0001-75   
2                      CEFOTAXIMA SÓDICA  04.301.884/0001-75   
3           CLORIDRATO DE CIPROFLOXACINO  04.301.884/0001-75   
4                CEFALEXINA MONOIDRATADA  04.301.884/0001-75   
...                                  ...                 ...   
27192  ACETATO DE RETINOL;COLECALCIFEROL  92.265.552/0001-40   
27193            ACETATO DE DEXAMETASONA  92.265.552/0001-40   
27194            ACETATO DE DEXAMETASONA  92.265.552/0001-40   
27195            ACETATO DE DEXAMETASONA  92.265.552/0001-40   
27196            ACETATO DE DEXAMETASONA  92.265.552/0001-40   

                                             LABORATÓRIO     CÓDIGO GGREM  \
0       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100106155118   
1       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100204157115   
2       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100310151117   
3       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100410113111   
4       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100801112111   
...                                                  ...              ...   
27192  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525917060050504   
27193  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525921120068107   
27194  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525921120067907   
27195  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525922030071603   
27196  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525922030071703   

            REGISTRO          EAN 1          EAN 2       EAN 3  \
0      1516700060050  7898361881450         -           -        
1      1516700030038  7898361881405         -           -        
2      1516700050063  7898361881412         -           -        
3      1516700070099  7898361881313         -           -        
4      1516700080035  7898361880019         -           -        
...              ...            ...            ...         ...   
27192  1181901840031  7896472518821         -           -        
27193  1181902220010  7896472503599  7895296449267      -        
27194  1181902240011  7896472502110  7896004782560      -        
27195  1181902240021  7896004782225         -           -        
27196  1181902240038  7896004782232         -           -        

                            PRODUTO  \
0                 CEFALOTINA SÓDICA   
1                 CEFAZOLINA SÓDICA   
2                 CEFOTAXIMA SÓDICA   
3      CLORIDRATO DE CIPROFLOXACINO   
4                        CEFALEXINA   
...                             ...   
27192                       METOSAD   
27193       ACETATO DE DEXAMETASONA   
27194                      CORTITOP   
27195                      CORTITOP   
27196                      CORTITOP   

                                            APRESENTAÇÃO  ... PMC 18% ALC  \
0                        1G PO SOL INJ CT 50 FA VD TRANS  ...         NaN   
1                       1 G PO SOL INJ CT 50 FA VD TRANS  ...         NaN   
2                       1 G PO SOL INJ CT 50 FA VD TRANS  ...         NaN   
3              500 MG COM REV CT 2 BL AL PLAS TRANS X 07  ...       46,26   
4           500 MG COM REV CT BL AL PVC/PVDC TRANS X 8    ...       20,97   
...                                                  ...  ...         ...   
27192  50000 + 10000 UI/ML SOL OR CT FR PLAS AMB GOT ...  ...       27,91   
27193                 1 MG/G CREM DERM CT BG AL X 10 G    ...       12,30   
27194                 1 MG/G CREM DERM CT BG AL X 10 G    ...       19,05   
27195                   1 MG/G CREM DERM CT BG AL X 20 G  ...       37,92   
27196                   1 MG/G CREM DERM CT BG AL X 30 G  ...       56,86   

      PMC 20% RESTRIÇÃO HOSPITALAR  CAP CONFAZ 87 ICMS 0% ANÁLISE RECURSAL  \
0         NaN                  Sim  Não       Não     Não              

##Tratativa dos dados

Conforme visto na base de dados, a coluna da classe terapêutica pode ter seus valores resumidos e unificados. Isso porque existe mais de uma descrição de classe terapêutica para um mesmo produto, e essa unificação nos ajuda a extrair informações interessantes.
<br/>Abaixo, foram selecionadas as classes terapêuticas que mais se repetiam e que possuiam mais de uma descrição. Vale ressaltar que nem todas foram mapeadas, mas apenas as principais ou mais comuns, sendo essas categorizadas como "Outro".

In [124]:
#df['CLASSE TERAPÊUTICA'].unique()

In [101]:
def map_classe_ter(classe):
  #print(classe)
  if(classe.find("ANTISSEPTICOS") != -1):
    return "ANTISSEPTICOS"
  if(classe.find("HORMÔNIOS") != -1 or classe.find("TESTOSTERONA") != -1 or classe.find("GONADOTROFINAS") != -1 or classe.find("ESTRÓGENOS") != -1):
    return "HORMÔNIOS"
  if(classe.find("ANTIVIRAIS") != -1 or classe.find("VIRAIS") != -1):
    return "ANTIVIRAIS"
  if(classe.find("SOLUÇÕES") != -1):
    return "SOLUÇÕES"
  if(classe.find("ANESTÉSICOS") != -1):
    return "ANESTÉSICOS"
  if(classe.find("INSULINA") != -1):
    return "INSULINA"
  if(classe.find("ANTIPSICÓTICOS") != -1):
    return "ANTIPSICÓTICOS"
  if(classe.find("Vacinas") != -1):
    return "VACINA"
  if(classe.find("ANTIDIABÉTICOS") != -1 or classe.find("DIABETES") != -1):
    return "ANTIDIABÉTICOS"
  if(classe.find("ANTICORPOS") != -1):
    return "ANTICORPOS"
  if(classe.find("ANTIFÚNGICOS") != -1):
    return "ANTIFÚNGICOS"
  if(classe.find("ANTITUSSÍGENOS") != -1):
    return "ANTITUSSÍGENOS"
  if(classe.find("CEFALOSPORINAS") != -1):
    return "CEFALOSPORINAS"
  if(classe.find("ANTIULCEROSOS") != -1):
    return "ANTIULCEROSOS"
  if(classe.find("SORO") != -1):
    return "SORO"
  if(classe.find("ANTINEOPLÁSICOS") != -1):
    return "ANTINEOPLÁSICOS"
  if(classe.find("ANTIBIÓTICOS") != -1):
    return "ANTIBIÓTICOS"
  if(classe.find("LAXANTES") != -1):
    return "LAXANTES"
  if(classe.find("ANTIÁCIDOS") != -1):
    return "ANTIACIDO"
  if(classe.find("ANTICONCEPCIONAIS") != -1):
    return "ANTICONCEPCIONAIS"
  if(classe.find("CORTICOSTERÓIDES") != -1):
    return "CORTICOSTERÓIDES "
  if(classe.find("ANTIULCEROSOS") != -1):
    return "ANTIULCEROSOS"
  if(classe.find("ANTIDEPRESSIVOS") != -1):
    return "ANTIDEPRESSIVOS"
  if(classe.find("ANTIASMÁTICOS") != -1):
    return "ANTIASMÁTICOS"
  if(classe.find("ANTI-HIPERTENSIVOS") != -1):
    return "ANTI-HIPERTENSIVOS"
  if(classe.find("ANALGÉSICOS") != -1):
    return "ANALGÉSICOS"
  if(classe.find("POLIVITAMÍNICO") != -1 or classe.find("VITAMINA") != -1):
    return "VITAMINA"
  if(classe.find("FITOTERÁPICOS") != -1):
    return "FITOTERÁPICOS"
  if(classe.find("INIBIDORES") != -1):
    return "INIBIDORES"
  if(classe.find("ANTIALÉRGICOS") != -1):
    return "ANTIALÉRGICOS"
  if(classe.find("ANTIINFLAMATÓRIOS") != -1):
    return "ANTIINFLAMATÓRIOS"
  if(classe.find("PENICILINAS") != -1):
    return "PENICILINAS"
  else:
    return "Outro"
df['CLASSE TERAPÊUTICA'] = df['CLASSE TERAPÊUTICA'].apply(lambda x: map_classe_ter(x))
df['CLASSE TERAPÊUTICA'].unique()

array(['CEFALOSPORINAS', 'Outro', 'ANTIDIABÉTICOS', 'PENICILINAS',
       'ANTIDEPRESSIVOS', 'HORMÔNIOS', 'INIBIDORES', 'ANTIPSICÓTICOS',
       'ANESTÉSICOS', 'ANTI-HIPERTENSIVOS', 'ANALGÉSICOS',
       'ANTIASMÁTICOS', 'LAXANTES', 'ANTITUSSÍGENOS', 'VITAMINA',
       'ANTIFÚNGICOS', 'ANTIULCEROSOS', 'CORTICOSTERÓIDES ',
       'ANTIALÉRGICOS', 'ANTIVIRAIS', 'ANTIINFLAMATÓRIOS', 'ANTIBIÓTICOS',
       'ANTIACIDO', 'ANTICORPOS', 'ANTINEOPLÁSICOS', 'SOLUÇÕES',
       'FITOTERÁPICOS', 'VACINA', 'INSULINA', 'ANTISSEPTICOS', 'SORO'],
      dtype=object)

As duas células seguintes removem algumas colunas inutilizadas e transforma os valores formatados em String para float, o que é necessário para avaliações mais a frente.

In [102]:
df.drop('EAN 2', inplace=True, axis=1)
df.drop('EAN 3', inplace=True, axis=1)
df.drop('ANÁLISE RECURSAL', inplace=True, axis=1)


In [103]:
df['PMC 17%'] = df['PMC 17%'].str.replace(',', '.')
df['PMC 18%'] = df['PMC 18%'].str.replace(',', '.')
df['PMC 20%'] = df['PMC 20%'].str.replace(',', '.')
df['PF 0%'] = df['PF 0%'].str.replace(',', '.')
df['PF Sem Impostos'] = df['PF Sem Impostos'].str.replace(',', '.')

df['PMC 17%'] = df['PMC 17%'].astype(float)
df['PMC 18%'] = df['PMC 18%'].astype(float)
df['PMC 20%'] = df['PMC 20%'].astype(float)
df['PF 0%'] = df['PF 0%'].astype(float)
df['PF Sem Impostos'] = df['PF Sem Impostos'].astype(float)

Neste ponto, assim como as classes terapêuticas, as tarjas também podem ser unificadas, uma vez que existe mais de uma descrição para uma mesma tarja. Além disso, as variações desse campo não são pertinentes para nossa análise e, portanto, podem ser unidas.

In [104]:
def map_tarjas(tarja):
  if tarja in ['Tarja Vermelha (*)','Tarja Vermelha(*)','Tarja  Vermelha', 'Tarja Vermelha', 'Tarja Vermelha sob restrição']:
    return 'Tarja Vermelha'
  if tarja in ['Tarja Preta(*)','Tarja  Preta']:
    return 'Tarja Preta'
  if tarja in ['Tarja Venda Livre/Sem Tarja (*)', 'Tarja Venda Livre', 'Tarja -(*)', '- (*) ','Tarja Sem Tarja']:
    return 'Sem Tarja'

df['TARJA'] = df['TARJA'].apply(map_tarjas)

#Regras de associação

Conforme proprosto inicialmente, temos por objetivo a previsão de preço de medicamentos a partir de algumas variáveis como tarja, classe terapêutica, laboratório e substância.
<br/>Para que consigamos fazer as regras de associação, precisamos discretizar o valor dos remédios, ou seja, transformá-los de número com ponto flutuante para string (variável categórica). Vale ressaltar que escolhemos a coluna PF sem Impostos com o objetivo prever o preço sobre o seu valor bruto, sem acréscimos de ICMS.
<br/> Dessa forma, definimos, de forma arbitrária, os labels de cada faixa de preço, sendo eles:
*   **Barato:** preço até R\$100,00
*   **Médio:** preço até R\$500,00
*   **Caro:** preço até R\$1000,00
*   **Muito caro:** preço acima de R\$1000,00



In [105]:
def map_valor(valor):
  if(valor > 0 and valor <= 100):
    return "Barato"
  elif(valor > 100 and valor <= 500):
    return "Médio"
  elif(valor > 500 and valor <= 1000):
    return "Caro"
  elif (valor > 1000):
    return "Muito caro"

df['PF DISC'] = df['PF Sem Impostos'].apply(lambda x : map_valor(x))
df[['PF Sem Impostos', 'PF DISC']]

PF Sem Impostos     PF DISC
0               262.28       Médio
1               517.93        Caro
2              1917.29  Muito caro
3                27.44      Barato
4                12.44      Barato
...                ...         ...
27192            16.55      Barato
27193             7.30      Barato
27194            11.30      Barato
27195            22.49      Barato
27196            33.73      Barato

[27197 rows x 2 columns]

In [106]:
#Excluindo as linhas cujos atributos PF Sem Impostos ocasionou em NaN na coluna PF DISC
df = df[df['PF DISC'].notna()]
df['PF DISC'].unique()

array(['Médio', 'Caro', 'Muito caro', 'Barato'], dtype=object)

##Selecionando atributos relevantes

In [107]:
df.columns

Index(['SUBSTÂNCIA', 'CNPJ', 'LABORATÓRIO', 'CÓDIGO GGREM', 'REGISTRO',
       'EAN 1', 'PRODUTO', 'APRESENTAÇÃO', 'CLASSE TERAPÊUTICA',
       'TIPO DE PRODUTO (STATUS DO PRODUTO)', 'REGIME DE PREÇO',
       'PF Sem Impostos', 'PF 0%', 'PF 12%', 'PF 17%', 'PF 17% ALC',
       'PF 17,5%', 'PF 17,5% ALC', 'PF 18%', 'PF 18% ALC', 'PF 20%', 'PMC 0%',
       'PMC 12%', 'PMC 17%', 'PMC 17% ALC', 'PMC 17,5%', 'PMC 17,5% ALC',
       'PMC 18%', 'PMC 18% ALC', 'PMC 20%', 'RESTRIÇÃO HOSPITALAR', 'CAP',
       'CONFAZ 87', 'ICMS 0%',
       'LISTA DE CONCESSÃO DE CRÉDITO TRIBUTÁRIO (PIS/COFINS)',
       'COMERCIALIZAÇÃO 2020', 'TARJA', 'PF DISC'],
      dtype='object')

In [108]:
df_rules = df[['SUBSTÂNCIA','LABORATÓRIO','CLASSE TERAPÊUTICA', 'TARJA','PF DISC']]
df_rules.head()

SUBSTÂNCIA  \
0             CEFALOTINA SODICA   
1             CEFAZOLINA SÓDICA   
2             CEFOTAXIMA SÓDICA   
3  CLORIDRATO DE CIPROFLOXACINO   
4       CEFALEXINA MONOIDRATADA   

                                        LABORATÓRIO CLASSE TERAPÊUTICA  \
0  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA     CEFALOSPORINAS   
1  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA     CEFALOSPORINAS   
2  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA     CEFALOSPORINAS   
3  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA              Outro   
4  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA     CEFALOSPORINAS   

            TARJA     PF DISC  
0  Tarja Vermelha       Médio  
1  Tarja Vermelha        Caro  
2  Tarja Vermelha  Muito caro  
3  Tarja Vermelha      Barato  
4  Tarja Vermelha      Barato

In [109]:
df_transactions = pd.get_dummies(df_rules)

In [110]:
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27190 entries, 0 to 27196
Columns: 2630 entries, SUBSTÂNCIA_21-ACETATO DE DEXAMETASONA;CLOTRIMAZOL to PF DISC_Médio
dtypes: uint8(2630)
memory usage: 68.4 MB


##Extraindo itemsets frequentes

Para a extração dos itemsets frequentes levamos em consideração um suporte mínimo de 0,01 (ou 1%). Isso porque grande parte das ocorrências tiveram suporte muito pequeno, e qualquer aumento no parâmetro trazia apenas algumas dezenas de transações.

In [132]:
frequent_itemsets = apriori(df_transactions, min_support = 0.01, use_colnames=True)
frequent_itemsets

support                                           itemsets
0    0.017654                        (SUBSTÂNCIA_COLECALCIFEROL)
1    0.011217                          (SUBSTÂNCIA_RIVAROXABANA)
2    0.010739             (LABORATÓRIO_ACCORD FARMACÊUTICA LTDA)
3    0.043509  (LABORATÓRIO_ACHÉ LABORATÓRIOS FARMACÊUTICOS S.A)
4    0.014049   (LABORATÓRIO_ALTHAIA S.A INDÚSTRIA FARMACÊUTICA)
..        ...                                                ...
213  0.019051  (CLASSE TERAPÊUTICA_Outro, LABORATÓRIO_EMS S/A...
214  0.011438  (LABORATÓRIO_EUROFARMA LABORATÓRIOS S.A., CLAS...
215  0.011512  (LABORATÓRIO_GERMED FARMACEUTICA LTDA, CLASSE ...
216  0.011879  (LABORATÓRIO_LEGRAND PHARMA INDÚSTRIA FARMACÊU...
217  0.010445  (LABORATÓRIO_NOVA QUIMICA FARMACÊUTICA S/A, CL...

[218 rows x 2 columns]

##Extraindo regras de associação

Neste momento extraimos as regras de associação a partir dos itemsets frequentes. Importante salientar que definimos um limite mínimo de 20% para a confiança para que mais combinações fossem trazidas.

In [143]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.2)
rules.sort_values(by="confidence", ascending=False)

antecedents  \
115  (SUBSTÂNCIA_COLECALCIFEROL, TARJA_Tarja Vermelha)   
15                (LABORATÓRIO_BLAU FARMACÊUTICA S.A.)   
56   (LABORATÓRIO_SANDOZ DO BRASIL INDÚSTRIA FARMAC...   
62                (LABORATÓRIO_TORRENT DO BRASIL LTDA)   
198  (LABORATÓRIO_SANDOZ DO BRASIL INDÚSTRIA FARMAC...   
..                                                 ...   
28                               (LABORATÓRIO_EMS S/A)   
240         (CLASSE TERAPÊUTICA_Outro, PF DISC_Barato)   
104                                   (PF DISC_Barato)   
149                              (LABORATÓRIO_EMS S/A)   
260   (CLASSE TERAPÊUTICA_Outro, TARJA_Tarja Vermelha)   

                               consequents  antecedent support  \
115          (CLASSE TERAPÊUTICA_VITAMINA)            0.014454   
15                  (TARJA_Tarja Vermelha)            0.013534   
56                  (TARJA_Tarja Vermelha)            0.013350   
62                  (TARJA_Tarja Vermelha)            0.012836   
198                 (TARJA_Tarja Vermelha)            0.010261   
..                                     ...                 ...   
28                         (PF DISC_Médio)            0.057889   
240                      (TARJA_Sem Tarja)            0.340051   
104                      (TARJA_Sem Tarja)            0.619051   
149  (PF DISC_Médio, TARJA_Tarja Vermelha)            0.057889   
260                        (PF DISC_Médio)            0.399448   

     consequent support   support  confidence       lift  leverage  conviction  
115            0.027584  0.014454    1.000000  36.253333  0.014055         inf  
15             0.806473  0.013387    0.989130   1.226489  0.002472   17.804487  
56             0.806473  0.013203    0.988981   1.226304  0.002437   17.562578  
62             0.806473  0.012652    0.985673   1.222203  0.002300   13.508187  
198            0.806473  0.010114    0.985663   1.222190  0.001839   13.498510  
..                  ...       ...         ...        ...       ...         ...  
28             0.234682  0.013203    0.228081   0.971874 -0.000382    0.991449  
240            0.173262  0.074770    0.219879   1.269053  0.015852    1.059756  
104            0.173262  0.135601    0.219047   1.264252  0.028343    1.058627  
149            0.204524  0.012210    0.210928   1.031311  0.000371    1.008116  
260            0.234682  0.082530    0.206611   0.880387 -0.011213    0.964619  

[312 rows x 9 columns]

###Extraindo métricas - Prevendo preço a partir da tarja

Aqui estimamos o preço de um medicamento tendo como base apenas a tarja, ou seja, dado uma classificação de tarja, qual o preço estimado apra aquele medicamento. Nisso, apresentamos o suporte, a confiança e lift para cada combinação de preço e tarja.

In [144]:
def calcula_metricas(att_antecedente, att_consequente):
  print("(",att_antecedente,") -> (",att_consequente,")")

  suporte = len(df_transactions[(df_transactions[att_antecedente] == 1) & (df_transactions[att_consequente] == 1) ])/ \
  len(df_transactions)
  print("Suporte: ",suporte)

  confianca = len(df_transactions[(df_transactions[att_antecedente] == 1) & (df_transactions[att_consequente] == 1) ])/ \
  len(df_transactions[df_transactions[att_antecedente] == 1])
  print("Confiança: ",confianca)

  lift = confianca / (len(df_transactions[df_transactions[att_consequente] == 1]) / len(df_transactions))

  print("Lift: ",lift)

  print("--------------------------------")
print("##########TARJA VERMELHA##########")
calcula_metricas('TARJA_Tarja Vermelha',"PF DISC_Barato")
calcula_metricas('TARJA_Tarja Vermelha',"PF DISC_Médio")
calcula_metricas('TARJA_Tarja Vermelha',"PF DISC_Caro")
calcula_metricas('TARJA_Tarja Vermelha',"PF DISC_Muito caro")
print("\n\n##########TARJA PRETA##########")
calcula_metricas('TARJA_Tarja Preta',"PF DISC_Barato")
calcula_metricas('TARJA_Tarja Preta',"PF DISC_Médio")
calcula_metricas('TARJA_Tarja Preta',"PF DISC_Caro")
calcula_metricas('TARJA_Tarja Preta',"PF DISC_Muito caro")
print("\n\n##########SEM TARJA##########")
calcula_metricas('TARJA_Sem Tarja',"PF DISC_Barato")
calcula_metricas('TARJA_Sem Tarja',"PF DISC_Médio")
calcula_metricas('TARJA_Sem Tarja',"PF DISC_Caro")
calcula_metricas('TARJA_Sem Tarja',"PF DISC_Muito caro")

##########TARJA VERMELHA##########
( TARJA_Tarja Vermelha ) -> ( PF DISC_Barato )
Suporte:  0.47043030525928653
Confiança:  0.5833181320685881
Lift:  0.9422778048327536
--------------------------------
( TARJA_Tarja Vermelha ) -> ( PF DISC_Médio )
Suporte:  0.20452372195660168
Confiança:  0.25360269974461874
Lift:  1.0806233201780573
--------------------------------
( TARJA_Tarja Vermelha ) -> ( PF DISC_Caro )
Suporte:  0.041780066200809124
Confiança:  0.051805910251732944
Lift:  1.094485392186961
--------------------------------
( TARJA_Tarja Vermelha ) -> ( PF DISC_Muito caro )
Suporte:  0.08973887458624494
Confiança:  0.1112732579350602
Lift:  1.1247285811354226
--------------------------------


##########TARJA PRETA##########
( TARJA_Tarja Preta ) -> ( PF DISC_Barato )
Suporte:  0.013019492460463406
Confiança:  0.6424682395644283
Lift:  1.0378274378420156
--------------------------------
( TARJA_Tarja Preta ) -> ( PF DISC_Médio )
Suporte:  0.005516734093416698
Confiança:  0.272232

###Obtendo combinações de antecendetes possíveis para cada preço

A regra com maior confiança (100%) é a que possui COLECALCIFEROL como substância e tarja vermelha.

In [114]:
rules[rules['confidence'] == rules.confidence.max()]

antecedents  \
115  (SUBSTÂNCIA_COLECALCIFEROL, TARJA_Tarja Vermelha)   

                       consequents  antecedent support  consequent support  \
115  (CLASSE TERAPÊUTICA_VITAMINA)            0.014454            0.027584   

      support  confidence       lift  leverage  conviction  
115  0.014454         1.0  36.253333  0.014055         inf

Aqui obtemos os antecedentes para que o preço classificado como barato seja alcançado.

In [150]:
rules[rules['consequents'].apply(lambda x: x.issubset(frozenset({'PF DISC_Barato'})))].sort_values(by="lift",ascending=False)

antecedents       consequents  \
19                   (LABORATÓRIO_CIMED INDUSTRIA S.A)  (PF DISC_Barato)   
132  (CLASSE TERAPÊUTICA_Outro, LABORATÓRIO_BRAINFA...  (PF DISC_Barato)   
75                   (CLASSE TERAPÊUTICA_ANTIFÚNGICOS)  (PF DISC_Barato)   
83              (CLASSE TERAPÊUTICA_CORTICOSTERÓIDES )  (PF DISC_Barato)   
46   (LABORATÓRIO_MULTILAB INDÚSTRIA E COMÉRCIO DE ...  (PF DISC_Barato)   
..                                                 ...               ...   
100                      (CLASSE TERAPÊUTICA_SOLUÇÕES)  (PF DISC_Barato)   
53            (LABORATÓRIO_PRATI DONADUZZI & CIA LTDA)  (PF DISC_Barato)   
196  (LABORATÓRIO_PRATI DONADUZZI & CIA LTDA, TARJA...  (PF DISC_Barato)   
23   (LABORATÓRIO_CRISTÁLIA PRODUTOS QUÍMICOS FARMA...  (PF DISC_Barato)   
79                 (CLASSE TERAPÊUTICA_ANTIPSICÓTICOS)  (PF DISC_Barato)   

     antecedent support  consequent support   support  confidence      lift  \
19             0.011512            0.619051  0.010519    0.913738  1.476030   
132            0.011401            0.619051  0.010298    0.903226  1.459049   
75             0.012431            0.619051  0.010960    0.881657  1.424207   
83             0.024752            0.619051  0.020743    0.838039  1.353747   
46             0.012284            0.619051  0.010261    0.835329  1.349371   
..                  ...                 ...       ...         ...       ...   
100            0.022950            0.619051  0.011070    0.482372  0.779212   
53             0.032917            0.619051  0.014527    0.441341  0.712931   
196            0.028761            0.619051  0.011916    0.414322  0.669286   
23             0.026444            0.619051  0.010188    0.385257  0.622335   
79             0.030048            0.619051  0.010592    0.352509  0.569435   

     leverage  conviction  
19   0.003392    4.416185  
132  0.003240    3.936472  
75   0.003264    3.219018  
83   0.005420    2.352097  
46   0.002657    2.313399  
..        ...         ...  
100 -0.003137    0.735951  
53  -0.005850    0.681898  
196 -0.005888    0.650441  
23  -0.006182    0.619688  
79  -0.008009    0.588346  

[72 rows x 9 columns]

Aqui obtemos os antecedentes para que o preço classificado como médio seja alcançado.

In [149]:
rules[rules['consequents'].apply(lambda x: x.issubset(frozenset({'PF DISC_Médio'})))].sort_values(by="lift",ascending=False)

antecedents      consequents  \
101                      (CLASSE TERAPÊUTICA_SOLUÇÕES)  (PF DISC_Médio)   
226  (CLASSE TERAPÊUTICA_ANTIPSICÓTICOS, TARJA_Tarj...  (PF DISC_Médio)   
80                 (CLASSE TERAPÊUTICA_ANTIPSICÓTICOS)  (PF DISC_Médio)   
24   (LABORATÓRIO_CRISTÁLIA PRODUTOS QUÍMICOS FARMA...  (PF DISC_Médio)   
72                (CLASSE TERAPÊUTICA_ANTIDEPRESSIVOS)  (PF DISC_Médio)   
217  (TARJA_Tarja Vermelha, CLASSE TERAPÊUTICA_ANTI...  (PF DISC_Médio)   
238  (TARJA_Tarja Vermelha, CLASSE TERAPÊUTICA_INIB...  (PF DISC_Médio)   
88                     (CLASSE TERAPÊUTICA_INIBIDORES)  (PF DISC_Médio)   
54            (LABORATÓRIO_PRATI DONADUZZI & CIA LTDA)  (PF DISC_Médio)   
147        (LABORATÓRIO_EMS S/A, TARJA_Tarja Vermelha)  (PF DISC_Médio)   
112                             (TARJA_Tarja Vermelha)  (PF DISC_Médio)   
10   (LABORATÓRIO_ACHÉ LABORATÓRIOS FARMACÊUTICOS S.A)  (PF DISC_Médio)   
68                    (CLASSE TERAPÊUTICA_ANALGÉSICOS)  (PF DISC_Médio)   
28                               (LABORATÓRIO_EMS S/A)  (PF DISC_Médio)   
260   (CLASSE TERAPÊUTICA_Outro, TARJA_Tarja Vermelha)  (PF DISC_Médio)   

     antecedent support  consequent support   support  confidence      lift  \
101            0.022950            0.234682  0.010298    0.448718  1.912026   
226            0.028099            0.234682  0.012026    0.428010  1.823790   
80             0.030048            0.234682  0.012688    0.422277  1.799358   
24             0.026444            0.234682  0.010739    0.406120  1.730511   
72             0.049982            0.234682  0.019603    0.392200  1.671199   
217            0.045862            0.234682  0.017470    0.380914  1.623109   
238            0.100809            0.234682  0.036925    0.366290  1.560792   
88             0.104965            0.234682  0.038396    0.365802  1.558716   
54             0.032917            0.234682  0.010004    0.303911  1.294990   
147            0.046267            0.234682  0.012210    0.263911  1.124548   
112            0.806473            0.234682  0.204524    0.253603  1.080623   
10             0.043509            0.234682  0.010776    0.247675  1.055367   
68             0.047738            0.234682  0.011328    0.237288  1.011106   
28             0.057889            0.234682  0.013203    0.228081  0.971874   
260            0.399448            0.234682  0.082530    0.206611  0.880387   

     leverage  conviction  
101  0.004912    1.388251  
226  0.005432    1.337993  
80   0.005637    1.324714  
24   0.004533    1.288674  
72   0.007873    1.259161  
217  0.006707    1.236207  
238  0.013267    1.207678  
88   0.013763    1.206750  
54   0.002279    1.099454  
147  0.001352    1.039709  
112  0.015259    1.025350  
10   0.000565    1.017271  
68   0.000124    1.003417  
28  -0.000382    0.991449  
260 -0.011213    0.964619

Aqui buscamos os antecedentes para que o preço classificado como caro seja alcançado. Porém, como é possível ver, não houve ocorrências, o que mostra que não foram encontradas condições para que um medicamento seja caro.

In [151]:
rules[rules['consequents'].apply(lambda x: x.issubset(frozenset({'PF DISC_Caro'})))].sort_values(by="lift",ascending=False)

Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, leverage, conviction]
Index: []

Por fim, obtemos os antecedentes para que o preço classificado como barato seja alcançado.

In [152]:
rules[rules['consequents'].apply(lambda x: x.issubset(frozenset({'PF DISC_Muito caro'})))].sort_values(by="lift",ascending=False)

antecedents           consequents  \
77                (CLASSE TERAPÊUTICA_ANTINEOPLÁSICOS)  (PF DISC_Muito caro)   
223  (CLASSE TERAPÊUTICA_ANTINEOPLÁSICOS, TARJA_Tar...  (PF DISC_Muito caro)   

     antecedent support  consequent support   support  confidence      lift  \
77             0.028908            0.098933  0.018242    0.631043  6.378463   
223            0.028356            0.098933  0.017874    0.630350  6.371458   

     leverage  conviction  
77   0.015382    2.442201  
223  0.015069    2.437622

###Estimando o preço do medicamento a partir da classe terapêutica

Abaixo vemos as regras de associação cujos antecedentes são as classes terapêuticas e a partir delas obtemos o valor.
<br/>Vendos os resultados, podemos tirar algumas conclusões:

*   Pode-se dizer com 41% de confiança e 35.22 pontos de lift, aproximadamente, que vitaminas são baratas;
*   Pode-se dizer com 63% de confiança e 6.38 pontos de lift, aproximadamente,que antineoplásticos são muito caros;
*   Pode-se dizer com 45% de confiança e 1.9 pontos de lift aproximadamente que soluções são baratas;
*   Pode-se dizer com 39% de confiança e 1.7 pontos de lift aproximadamente que antidepressivos possuem preço médico;
*   Pode-se dizer com 75% de confiança e 1.3 pontos de lift aproximadamente que antidiabéticos são baratos;
<br/><br/>
entre outras.


In [160]:
rules[rules['antecedents'].apply(lambda x: x.issubset(frozenset({'CLASSE TERAPÊUTICA_CEFALOSPORINAS', 'CLASSE TERAPÊUTICA_Outro', 'CLASSE TERAPÊUTICA_ANTIDIABÉTICOS', 'CLASSE TERAPÊUTICA_PENICILINAS',
       'CLASSE TERAPÊUTICA_ANTIDEPRESSIVOS', 'CLASSE TERAPÊUTICA_HORMÔNIOS', 'CLASSE TERAPÊUTICA_INIBIDORES', 'CLASSE TERAPÊUTICA_ANTIPSICÓTICOS',
       'CLASSE TERAPÊUTICA_ANESTÉSICOS', 'CLASSE TERAPÊUTICA_ANTI-HIPERTENSIVOS', 'CLASSE TERAPÊUTICA_ANALGÉSICOS',
       'CLASSE TERAPÊUTICA_ANTIASMÁTICOS', 'CLASSE TERAPÊUTICA_LAXANTES', 'CLASSE TERAPÊUTICA_ANTITUSSÍGENOS', 'CLASSE TERAPÊUTICA_VITAMINA',
       'CLASSE TERAPÊUTICA_ANTIFÚNGICOS', 'CLASSE TERAPÊUTICA_ANTIULCEROSOS', 'CLASSE TERAPÊUTICA_CORTICOSTERÓIDES ',
       'CLASSE TERAPÊUTICA_ANTIALÉRGICOS', 'CLASSE TERAPÊUTICA_ANTIVIRAIS', 'CLASSE TERAPÊUTICA_ANTIINFLAMATÓRIOS', 'CLASSE TERAPÊUTICA_ANTIBIÓTICOS',
       'CLASSE TERAPÊUTICA_ANTIACIDO', 'CLASSE TERAPÊUTICA_ANTICORPOS', 'CLASSE TERAPÊUTICA_ANTINEOPLÁSICOS', 'CLASSE TERAPÊUTICA_SOLUÇÕES',
       'CLASSE TERAPÊUTICA_FITOTERÁPICOS', 'CLASSE TERAPÊUTICA_VACINA', 'CLASSE TERAPÊUTICA_INSULINA', 'CLASSE TERAPÊUTICA_ANTISSEPTICOS', 'CLASSE TERAPÊUTICA_SORO'})))].sort_values(by="lift",ascending=False).reset_index()

index                             antecedents  \
0     116           (CLASSE TERAPÊUTICA_VITAMINA)   
1       1           (CLASSE TERAPÊUTICA_VITAMINA)   
2     122           (CLASSE TERAPÊUTICA_VITAMINA)   
3     224    (CLASSE TERAPÊUTICA_ANTINEOPLÁSICOS)   
4      77    (CLASSE TERAPÊUTICA_ANTINEOPLÁSICOS)   
5     212        (CLASSE TERAPÊUTICA_ANALGÉSICOS)   
6      65        (CLASSE TERAPÊUTICA_ANALGÉSICOS)   
7     227     (CLASSE TERAPÊUTICA_ANTIPSICÓTICOS)   
8     101           (CLASSE TERAPÊUTICA_SOLUÇÕES)   
9      80     (CLASSE TERAPÊUTICA_ANTIPSICÓTICOS)   
10    239         (CLASSE TERAPÊUTICA_INIBIDORES)   
11    218    (CLASSE TERAPÊUTICA_ANTIDEPRESSIVOS)   
12     72    (CLASSE TERAPÊUTICA_ANTIDEPRESSIVOS)   
13    221     (CLASSE TERAPÊUTICA_ANTIDIABÉTICOS)   
14    230  (CLASSE TERAPÊUTICA_CORTICOSTERÓIDES )   
15     88         (CLASSE TERAPÊUTICA_INIBIDORES)   
16     75       (CLASSE TERAPÊUTICA_ANTIFÚNGICOS)   
17     83  (CLASSE TERAPÊUTICA_CORTICOSTERÓIDES )   
18     74     (CLASSE TERAPÊUTICA_ANTIDIABÉTICOS)   
19    233          (CLASSE TERAPÊUTICA_HORMÔNIOS)   
20     76    (CLASSE TERAPÊUTICA_ANTINEOPLÁSICOS)   
21     98        (CLASSE TERAPÊUTICA_PENICILINAS)   
22     86         (CLASSE TERAPÊUTICA_INIBIDORES)   
23     73     (CLASSE TERAPÊUTICA_ANTIDIABÉTICOS)   
24     99           (CLASSE TERAPÊUTICA_SOLUÇÕES)   
25    103           (CLASSE TERAPÊUTICA_VITAMINA)   
26     69      (CLASSE TERAPÊUTICA_ANTIASMÁTICOS)   
27     67        (CLASSE TERAPÊUTICA_ANALGÉSICOS)   
28     78     (CLASSE TERAPÊUTICA_ANTIPSICÓTICOS)   
29     84          (CLASSE TERAPÊUTICA_HORMÔNIOS)   
30     82  (CLASSE TERAPÊUTICA_CORTICOSTERÓIDES )   
31     81         (CLASSE TERAPÊUTICA_ANTIVIRAIS)   
32     70    (CLASSE TERAPÊUTICA_ANTIDEPRESSIVOS)   
33    236         (CLASSE TERAPÊUTICA_INIBIDORES)   
34     93              (CLASSE TERAPÊUTICA_Outro)   
35    251              (CLASSE TERAPÊUTICA_Outro)   
36    215    (CLASSE TERAPÊUTICA_ANTIDEPRESSIVOS)   
37     68        (CLASSE TERAPÊUTICA_ANALGÉSICOS)   
38    266           (CLASSE TERAPÊUTICA_SOLUÇÕES)   
39    269           (CLASSE TERAPÊUTICA_VITAMINA)   
40     91              (CLASSE TERAPÊUTICA_Outro)   
41     85          (CLASSE TERAPÊUTICA_HORMÔNIOS)   
42     71    (CLASSE TERAPÊUTICA_ANTIDEPRESSIVOS)   
43     87         (CLASSE TERAPÊUTICA_INIBIDORES)   
44    102           (CLASSE TERAPÊUTICA_VITAMINA)   
45    100           (CLASSE TERAPÊUTICA_SOLUÇÕES)   
46     79     (CLASSE TERAPÊUTICA_ANTIPSICÓTICOS)   
47     66        (CLASSE TERAPÊUTICA_ANALGÉSICOS)   

                                          consequents  antecedent support  \
0   (SUBSTÂNCIA_COLECALCIFEROL, TARJA_Tarja Vermelha)            0.027584   
1                         (SUBSTÂNCIA_COLECALCIFEROL)            0.027584   
2         (SUBSTÂNCIA_COLECALCIFEROL, PF DISC_Barato)            0.027584   
3          (PF DISC_Muito caro, TARJA_Tarja Vermelha)            0.028908   
4                                (PF DISC_Muito caro)            0.028908   
5                   (PF DISC_Barato, TARJA_Sem Tarja)            0.047738   
6                                   (TARJA_Sem Tarja)            0.047738   
7               (PF DISC_Médio, TARJA_Tarja Vermelha)            0.030048   
8                                     (PF DISC_Médio)            0.022950   
9                                     (PF DISC_Médio)            0.030048   
10              (PF DISC_Médio, TARJA_Tarja Vermelha)            0.104965   
11              (PF DISC_Médio, TARJA_Tarja Vermelha)            0.049982   
12                                    (PF DISC_Médio)            0.049982   
13             (PF DISC_Barato, TARJA_Tarja Vermelha)            0.021405   
14             (PF DISC_Barato, TARJA_Tarja Vermelha)            0.024752   
15                                    (PF DISC_Médio)            0.104965   
16                                   (PF DISC_Barato)            0.012431   
17                  